challenge_storing_groceries/src/challenge_storing_groceries/open_door.py

```diff
@@ -0,0 +1,197 @@
+
+
+class OpenDoor(smach.State):
+    def __init__(self, robot, cabinet):
+        smach.State.__init__(self, outcomes=['succeeded', 'failed'])
+        self.robot = robot
+        self.cabinet = cabinet
+
+        self._rate = rospy.Rate(10)
+        self._goal_position_tolerance = 0.01
+        self._goal_rotation_tolerance = 0.1
+
+        self._tf_buffer = tf2_ros.Buffer()
+        self._tf_listener = tf2_ros.TransformListener(self._tf_buffer)
+        self._cmd_vel_publisher = rospy.Publisher("/" + robot.robot_name + "/base/references", Twist, queue_size=1)
+
+    def _get_target_delta_in_robot_frame(self, goal_pose):
+        goal_pose.header.stamp = rospy.Time.now()
+        pose = self._tf_buffer.transform(goal_pose, self.robot.robot_name + '/base_link', rospy.Duration(1.0))
+        yaw = _get_yaw_from_quaternion_msg(pose.pose.orientation)
+        return pose.pose.position.x, pose.pose.position.y, _wrap_angle_pi(yaw)
+
+    def _goal_reached(self, dx, dy, dyaw):
+        return math.hypot(dx, dy) < self._goal_position_tolerance and abs(dyaw) < self._goal_rotation_tolerance
+
+    def _control_to_pose(self, goal_pose, position_gain, rotation_gain, abs_vx, abs_vy, abs_vyaw):
+        if self._goal_reached(*self._get_target_delta_in_robot_frame(goal_pose)):
+            rospy.loginfo("We are already there")
+            return
+
+        rospy.loginfo("Starting alignment ....")
+        while not rospy.is_shutdown():
+            dx, dy, dyaw = self._get_target_delta_in_robot_frame(goal_pose)
+
+            if self._goal_reached(dx, dy, dyaw):
+                break
+
+            rospy.loginfo_throttle(1.0, "Aligning .. Delta = {} {} {}".format(dx, dy, dyaw))
+
+            self._cmd_vel_publisher.publish(Twist(
+                linear=Vector3(
+                    x=_clamp(abs_vx, position_gain * dx),
+                    y=_clamp(abs_vy, position_gain * dy)
+                ),
+                angular=Vector3(z=_clamp(abs_vyaw, rotation_gain * dyaw))
+            ))
+
+            self._rate.sleep()
+
+        rospy.loginfo("Goal reached")
+
+    def _align_with_cabinet(self):
+        goal_pose = PoseStamped()
+        goal_pose.header.stamp = rospy.Time.now()
+        goal_pose.header.frame_id = self.cabinet.id_
+        goal_pose.pose.orientation = Quaternion(*quaternion_from_euler(0, 0, math.pi - 0.05))
+        goal_pose.pose.position.x = 0.6
+        self._control_to_pose(goal_pose, 0.5, 1.0, 0.3, 0.3, 0.3)
+
+    def _return_from_cabinet(self):
+        goal_pose = PoseStamped()
+        goal_pose.header.stamp = rospy.Time.now()
+        goal_pose.header.frame_id = self.cabinet.id_
+        goal_pose.pose.orientation = Quaternion(*quaternion_from_euler(0, 0, math.pi))
+        goal_pose.pose.position.x = 1.0
+        self._control_to_pose(goal_pose, 0.5, 1.0, 0.5, 0.5, 0.5)
+
+    def _move_arm_in_cabinet(self):
+        self.robot.leftArm._send_joint_trajectory([[-0.101, 0.119, 0.200, 1.363, 0.230, 0.688, 0.280]])
```

Comment:
Please don't borrow specific arms of the robot, as they may disappear when the code is used at another robot. Eventually, we should ask the robot for an appropriate arm. As a first step, introduce a `self.arm` variable, assign the `robot.leftArm` to it, and use the former everywhere (assuming one arm is sufficient).

Also, `_send_joint_trajectory` is a private method, don't use it.

Reply:
Good point. This whole door opening is heavily tied to a very specific arm (Amigo's left arm) and very specif cabinet. 
I'm not sure its even worth including this PR at all, because it's so specific. 

I can move the trajectories to amigo_description skills.yaml and use theme here, that would clean up a lot.

Reply:
But since these joint goals are very specific to this one cabinet, I'm not going to take that effort. I will make the ```self.arm``` though.

Reply:
Removing the use of the hardcoded joint goals is https://github.com/tue-robotics/tue_robocup/issues/685